src/reducers/console.js

```diff
@@ -35,20 +33,62 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
+      return payload.trim(' ') === '' ? state : state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({expression: payload}),
       );
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
-        new ConsoleEntry({
+      return state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
+      );
+
+      const newHistoryEntryIndex = payload.direction === 'UP' ?
+        state.historyEntryIndex + 1 :
+        state.historyEntryIndex - 1;
+
+      if (newHistoryEntryIndex < 0 ||
+          newHistoryEntryIndex === relevantHistory.size) {
+        return state;
+      }
+
+      const downToPreviousInput = newHistoryEntryIndex === 0 &&
+        payload.direction === 'DOWN';
```

Comment:
I don’t think there’s any particular need to check this condition explicitly—the boundary guards just above should make this redundant.

Reply:
This is handling the user pressing down to get their previously typed input, so  it's not like the no-op out of bounds boundary cases handled above

Reply:
Sorry, just to be clear—I don’t think the second part of the conjunction (`payload.direction === 'DOWN'`) is needed; the only way to get to `0` is down, right?

Reply:
Oh I see, you're referring to the `=== 'DOWN'`. Right you are.